core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -505,14 +506,13 @@ private[kafka] class Processor(val id: Int,
     selector.completedReceives.asScala.foreach { receive =>
       try {
         val openChannel = selector.channel(receive.source)
-        val session = {
-          // Only methods that are safe to call on a disconnected channel should be invoked on 'channel'.
-          val channel = if (openChannel != null) openChannel else selector.closingChannel(receive.source)
-          RequestChannel.Session(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, channel.principal.getName), channel.socketAddress)
-        }
+        // Only methods that are safe to call on a disconnected channel should be invoked on 'openOrClosingChannel'.
+        val openOrClosingChannel = if (openChannel != null) openChannel else selector.closingChannel(receive.source)
+        val session = RequestChannel.Session(new KafkaPrincipal(KafkaPrincipal.USER_TYPE, openOrClosingChannel.principal.getName), openOrClosingChannel.socketAddress)
+        val networkThreadTimeNanos = openOrClosingChannel.getAndResetNetworkThreadTimeNanos
         val req = RequestChannel.Request(processor = id, connectionId = receive.source, session = session,
-          buffer = receive.payload, startTimeMs = time.milliseconds, listenerName = listenerName,
-          securityProtocol = securityProtocol)
+          buffer = receive.payload, startTimeNanos = time.nanoseconds, networkThreadTimeNanos = networkThreadTimeNanos,
+          listenerName = listenerName, securityProtocol = securityProtocol)
```

Comment:
After writing the comment above, I realized that it sounds rather odd. So I have updated the code to record network thread time when request metrics are updated, so that receive+send are recorded together. This does also ensure that the last network time on each connection is recorded and it would work even if requests on a connection use different clientIds. Throttling is still performed only on subsequent requests.